// ng-firebase-auth/package/src/directive.js
'use strict';

/**
 * @ngInject
 * @ngdoc object
 * @name app.directives:Authentication
 * @description Directive for authenticating users
 * @param  {Object} Firebase
 * @param  {Object} $templateCache
 * @param  {Object} $rootScope
 */
module.exports = function(Firebase, $templateCache, $rootScope) {

  return {

    restrict: 'E',

    scope: {
      firebase: '='
    },

    template: [
      '<div class="ng-firebase-auth flex-col center">',
        '<div class="message">{{message}}</div>',
        '<!-- Facebook Button -->',
        '<button class="facebook" ng-click="vm.onAuthButtonClick(\'facebook\')">',
          'Sign in with Facebook',
        '</button>',
        '<!-- Google Button -->',
        '<button class="google" ng-click="vm.onAuthButtonClick(\'google\')">',
          'Sign in with Google',
        '</button>',
        '<!-- Github Button -->',
        '<button class="github" ng-click="vm.onAuthButtonClick(\'github\')">',
          'Sign in with Github',
        '</button>',
      '</div>'
    ].join(''),

    controller: 'AuthCtrl',

    controllerAs: 'vm'

  };

};
